pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,706 @@
+	ipFlow.IsTimeOut = true
+
+	// Assume the App has an assigned IP address(es) first
+	if len(instData.appIPAddrs[AppNum]) > 0 {
+		ipFlow.dbg1 = 1
+		forwSrcApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Forward.SrcIP)
+		if forwSrcApp == false {
+			ipFlow.dbg1 = 2
+			forwDstApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Forward.DstIP)
+			if forwDstApp == false {
+				ipFlow.dbg1 = 3
+				backSrcApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Reverse.SrcIP)
+				if backSrcApp == false {
+					ipFlow.dbg1 = 4
+					backDstApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Reverse.DstIP)
+				} else {
+					ipFlow.dbg1 = 5 // XXX
+				}
+			}
+		}
+	}
+
+	if forwSrcApp {
+		ipFlow.dbg2 = 1
+		ipFlow.SrcIP = entry.Forward.SrcIP
+		ipFlow.DstIP = entry.Reverse.SrcIP
+		ipFlow.SrcPort = entry.Forward.SrcPort
+		ipFlow.DstPort = entry.Reverse.DstPort
+		ipFlow.AppInitiate = true
+	} else if forwDstApp {
+		ipFlow.dbg2 = 2
+		ipFlow.DstIP = entry.Forward.DstIP
+		ipFlow.SrcIP = entry.Reverse.DstIP
+		ipFlow.DstPort = entry.Forward.DstPort
+		ipFlow.SrcPort = entry.Reverse.DstPort
+	} else if backSrcApp {
+		ipFlow.dbg2 = 3
+		ipFlow.SrcIP = entry.Forward.SrcIP
+		ipFlow.DstIP = entry.Reverse.SrcIP
+		ipFlow.SrcPort = entry.Forward.SrcPort
+		ipFlow.DstPort = entry.Reverse.SrcPort
+	} else if backDstApp {
+		ipFlow.dbg2 = 4
+		ipFlow.SrcIP = entry.Reverse.DstIP
+		ipFlow.DstIP = entry.Forward.DstIP
+		ipFlow.SrcPort = entry.Reverse.DstPort
+		ipFlow.DstPort = entry.Forward.DstPort
+		ipFlow.AppInitiate = true
+	} else { // if we can not find our App endpoint is part of the flow, something is wrong
+		ipFlow.dbg2 = 5
+		if AppNum > 0 {
+			log.Infof("FlowStats: flow entry can not locate app IP address, appNum %d, %s", AppNum, entry.String())
+		}
+		return ipFlow
+	}
+
+	if ipFlow.AppInitiate {
+		ipFlow.SendPkts = entry.Forward.Packets
+		ipFlow.SendBytes = entry.Forward.Bytes
+		ipFlow.RecvPkts = entry.Reverse.Packets
+		ipFlow.RecvBytes = entry.Reverse.Bytes
+	} else {
+		ipFlow.SendPkts = entry.Reverse.Packets
+		ipFlow.SendBytes = entry.Reverse.Bytes
+		ipFlow.RecvPkts = entry.Forward.Packets
+		ipFlow.RecvBytes = entry.Forward.Bytes
+	}
+	return ipFlow
+}
+
+func checkAppIPAddr(appAddrs []net.IP, entryIP net.IP) bool {
+	for _, appIP := range appAddrs {
+		if strings.Compare(appIP.String(), entryIP.String()) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// print the FlowTuple entries
+func (s *flowStats) String() string {
+	tstart := time.Unix(0, int64(s.TimeStart))
+	tout := int32(s.TimeOut)
+	return fmt.Sprintf("TS %v, TO %d(sec), proto %d src=%s dst=%s sport=%d dport=%d, snd=pkts/bytes %d/%d rcv=pkts/bytes %d/%d app-init %v, appNum %d, aclnum %d",
+		tstart, tout, s.Proto, s.SrcIP, s.DstIP, s.SrcPort, s.DstPort, s.SendPkts, s.SendBytes,
+		s.RecvPkts, s.RecvBytes, s.AppInitiate, s.appNum, s.aclNum)
+}
+
+func checkAppAndACL(ctx *zedrouterContext, instData networkAttrs) {
```

Comment:
Not sure if I'm missing something here, but instData is passed by value here and updated inside this function. Caller of this function will not see the updates done inside here. This function works on a copy only.

Reply:
I had the same doubt, but i tried this and it works. maybe it's because they are the structures off this instData. It works, since the debug printout shows the data of ACL is filled inside this function correctly

Reply:
I have verified that: the simple type in a structure will not be changed inside a function by passing as value of the structure; but the array, map, slice inside the structure, since they are the pointers to those types, passing into functions will be changed inside those complex types. all my items in instData are complex types, basically pointers to those data areas, that is ok.